crates/ide/src/prime_caches.rs

```diff
@@ -3,10 +3,65 @@
 //! request takes longer to compute. This modules implemented prepopulating of
 //! various caches, it's not really advanced at the moment.
 
-use crate::{FileId, RootDatabase};
+use std::sync::atomic::AtomicUsize;
+use std::sync::atomic::Ordering;
 
-pub(crate) fn prime_caches(db: &RootDatabase, files: Vec<FileId>) {
-    for file in files {
-        let _ = crate::syntax_highlighting::highlight(db, file, None, false);
-    }
+use base_db::salsa::ParallelDatabase;
+use base_db::SourceDatabase;
+use hir::db::DefDatabase;
+
+use crate::RootDatabase;
+
+#[derive(Debug)]
+pub enum PrimeCachesProgress {
+    Started,
+    /// We started indexing a crate.
+    StartedOnCrate {
+        on_crate: String,
+        n_done: usize,
+        n_total: usize,
+    },
+    /// We finished indexing all crates.
+    Finished,
+}
+
+pub(crate) fn prime_caches<F>(db: &RootDatabase, cb: F)
+where
+    F: Fn(PrimeCachesProgress) + Sync,
+{
+    let _p = profile::span("prime_caches");
+    let graph = db.crate_graph();
+    let topo = &graph.crates_in_topological_order();
+
+    let cb = &cb;
+    let finished = &AtomicUsize::new(0);
+    let db = db.snapshot();
+
+    cb(PrimeCachesProgress::Started);
+
+    rayon::scope(move |s| {
```

Comment:
Hm, will this handle cancellation? I have a suspicion that no, it won't...

Reply:
Hm, docs imply that this should be OK: https://docs.rs/rayon/1.4.1/rayon/fn.scope.html#panics

But it seems at least plausible that this messes up on_typing test...